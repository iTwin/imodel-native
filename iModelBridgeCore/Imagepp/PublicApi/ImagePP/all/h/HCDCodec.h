//:>--------------------------------------------------------------------------------------+
//:>
//:>     $Source: PublicApi/ImagePP/all/h/HCDCodec.h $
//:>
//:>  $Copyright: (c) 2015 Bentley Systems, Incorporated. All rights reserved. $
//:>
//:>+--------------------------------------------------------------------------------------
//-----------------------------------------------------------------------------
// Class : HCDCodec
//-----------------------------------------------------------------------------
// General class for codecs.
//-----------------------------------------------------------------------------

#pragma once

#include <Imagepp/all/h/HFCExclusiveKey.h>

BEGIN_IMAGEPP_NAMESPACE

class HCDPacketRLE;

// Binary optimization: Speed up read/write when codec support direct decompression/compression to/from RLE format.
class HNOVTABLEINIT HCDCodecRLEInterface
    {
public:
    virtual void     DecompressSubsetToRLE(const void* pi_pInData, size_t pi_InDataSize, HFCPtr<HCDPacketRLE>& pio_rpRLEPacket) = 0;
    virtual size_t   CompressSubsetFromRLE(HFCPtr<HCDPacketRLE> const& pi_rpPacketRLE, void* po_pOutBuffer, size_t po_OutBufferSize) = 0;
    };

class HNOVTABLEINIT HCDCodec : public HFCShareableObject<HCDCodec>,
    public HFCExclusiveKey     // to have access to claim/release instead of including m_Key and adding the methods
    {
    HDECLARE_BASECLASS_ID(HCDCodecId_Base)

public:

    // codec states enumeration
    enum State
        {
        STATE_NONE,
        STATE_COMPRESS,
        STATE_DECOMPRESS
        };

    // destructor
    virtual         ~HCDCodec();

    // compression
    virtual size_t          CompressSubset(const void* pi_pInData,
                                           size_t pi_InDataSize,
                                           void* po_pOutBuffer,
                                           size_t pi_OutBufferSize) = 0;

    virtual size_t          DecompressSubset(const void* pi_pInData,
                                             size_t pi_InDataSize,
                                             void* po_pOutBuffer,
                                             size_t pi_OutBufferSize) = 0;

    State                   GetCurrentState() const;
    virtual void            Reset();

    virtual HCDCodec*       Clone() const = 0;

    virtual size_t          GetDataSize() const = 0;

    // subset
    virtual size_t          GetMinimumSubsetSize() const = 0;
    virtual size_t          GetSubsetMaxCompressedSize() const = 0;
    virtual void            SetSubsetSize(size_t pi_Size) = 0;

    // others
    virtual double          GetEstimatedCompressionRatio() const;
    virtual bool            HasRandomAccess() const;
    virtual bool            IsCodecImage() const;
    virtual bool            IsCodecVector() const;
    virtual const WString&  GetName() const;

    virtual HCDCodecRLEInterface* 
                            GetRLEInterface();

protected:

    // constructors
    HCDCodec(const WString& pi_rCodecName);
    HCDCodec(const HCDCodec& pi_rObj);


    // compression
    void                    SetCurrentState(State pi_State);

    WString         m_CodecName;

private:

    State          m_CurrentState;
    };

END_IMAGEPP_NAMESPACE