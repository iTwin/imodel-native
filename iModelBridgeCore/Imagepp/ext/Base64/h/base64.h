#ifndef _base64_h_
#define _base64_h_

#include <iostream>

namespace b64
{
        bool encode(std::istream& input, std::ostream& output);
        bool decode(std::istream& input, std::ostream& output);
};

#endif
