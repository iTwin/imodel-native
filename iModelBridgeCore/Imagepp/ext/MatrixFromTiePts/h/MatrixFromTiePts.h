/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See COPYRIGHT.md in the repository root for full copyright notice.
*--------------------------------------------------------------------------------------------*/

#pragma once


#define TRANSFO_NONE     -1 /* No transformation model */
#define TRANSFO_TRAN_2D   0 /* Translation en x et y */
#define TRANSFO_SIMI_2D   1 /* Translation x,y + Rotation + Echelle */
#define TRANSFO_POLY_2D_1 2 /* Polynomiale deg 1 */
#define TRANSFO_AFFI_2D   3 /* Tran. x,y + Rotation + Echelle + Axes */
#define TRANSFO_PROJ_2D   4 /* Projection d'un plan sur un autre */
#define TRANSFO_POLY_2D_2 5 /* Polynomiale deg 2 */
#define TRANSFO_POLY_2D_3 6 /* Polynomiale deg 3 */
#define TRANSFO_POLY_2D_4 7 /* Polynomiale deg 4 */ /* A tester ??? */
#define TRANSFO_POLY_2D_5 8 /* Polynomiale deg 5 */ /* A tester ??? */
#define TRANSFO_MAT_2D    9 /* Matrice de transformation */
#define TRANSFO_HELM_2D   10
#define TRANSFO_TPS_2D    11
#define TRANSFO_MAX       12

#define TRF_ERROR_MAX_ITERATIONS    -1
#define TRF_ERROR_NO_MEMORY         -2
#define TRF_ERROR_BAD_NUMBER_POINTS -3
#define TRF_ERROR_MATRIX            -4


long GetTransfoMatrixFromScaleAndTiePts (double          po_pMatrix[4][4],
                                         unsigned short  pi_NbVal_GeoTiePoint,
                                         double*         pi_pVal_GeoTiePoint,
                                         unsigned short  pi_NbVal_GeoPixelScale,
                                         double*         pi_pVal_GeoPixelScale);

long GetProjectiveTransfoMatrixFromScaleAndTiePts(double         po_pMatrix[4][4],
                                                  unsigned short pi_NbVal_GeoTiePoint,
                                                  double*        pi_pVal_GeoTiePoint);

long GetAffineTransfoMatrixFromScaleAndTiePts(double         po_pMatrix[4][4],
                                              unsigned short pi_NbVal_GeoTiePoint,
                                              double*        pi_pVal_GeoTiePoint);

long GetTransfoMatrixFromTiePts(double           po_pMatrix[4][4],
                                unsigned short   pi_NbVal_GeoTiePoint,
                                const double*    pi_pVal_GeoTiePoint,
                                long             pi_ModelType);

