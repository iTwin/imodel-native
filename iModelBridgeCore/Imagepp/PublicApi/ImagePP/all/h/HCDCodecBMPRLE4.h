//:>--------------------------------------------------------------------------------------+
//:>
//:>
//:>  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
//:>
//:>+--------------------------------------------------------------------------------------
// Class : HCDCodecBMPRLE8
//-----------------------------------------------------------------------------
// RLE8 codec lass.
//-----------------------------------------------------------------------------
#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecBMPRLE4 : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_BMPRLE4, HCDCodecImage)

public:

    IMAGEPP_EXPORT                 HCDCodecBMPRLE4();

    HCDCodecBMPRLE4(size_t pi_Width,
                    size_t pi_Height);

    HCDCodecBMPRLE4(const HCDCodecBMPRLE4& pi_rObj);

    ~HCDCodecBMPRLE4();

    size_t          CompressSubset(const void* pi_pInData,
                                   size_t     pi_InDataSize,
                                   void*      po_pOutBuffer,
                                   size_t     pi_OutBufferSize) override;

    size_t          DecompressSubset(const void* pi_pInData,
                                     size_t       pi_InDataSize,
                                     void*        po_pOutBuffer,
                                     size_t       pi_OutBufferSize) override;

    bool           HasLineAccess() const override;

    virtual HCDCodec* Clone() const override;

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const override;

    size_t          GetSubsetMaxCompressedSize() const override;

protected:

private:
    };

END_IMAGEPP_NAMESPACE
