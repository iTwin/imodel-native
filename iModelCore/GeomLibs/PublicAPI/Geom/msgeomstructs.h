/*--------------------------------------------------------------------------------------+
|
|     $Source: PublicAPI/Geom/msgeomstructs.h $
|
|  $Copyright: (c) 2011 Bentley Systems, Incorporated. All rights reserved. $
|
+--------------------------------------------------------------------------------------*/
#pragma once

/*__BENTLEY_INTERNAL_ONLY__*/

#include <Geom/GeomApi.h>


