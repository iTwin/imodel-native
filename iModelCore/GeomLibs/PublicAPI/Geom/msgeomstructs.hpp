/*--------------------------------------------------------------------------------------+
|
|     $Source: PublicAPI/Geom/msgeomstructs.hpp $
|
|  $Copyright: (c) 2009 Bentley Systems, Incorporated. All rights reserved. $
|
+--------------------------------------------------------------------------------------*/
#pragma once
/*__BENTLEY_INTERNAL_ONLY__*/
#include <Geom/GeomApi.h>
