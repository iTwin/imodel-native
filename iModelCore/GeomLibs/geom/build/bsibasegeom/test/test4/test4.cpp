/*--------------------------------------------------------------------------------------+
|
|  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
|
+--------------------------------------------------------------------------------------*/
// test1.cpp : Defines the entry point for the console application.
//

#include <stdio.h>
#include <msgeomstructs.hpp>
#include <msgeomstructs.h>
#include <printfuncs.h>
#include <stdlib.h>


#include <ellipseFixup.cpp>


static double s_pi = 4.0 * atan (1.0);

Public void showDPoint3d
(
char *pName,
DPoint3d const *pXYZ
)
    {
    printf ("    DPoint3d %s = {%18.11lf, %18.11lf, %18.11lf};\n",
            pName, pXYZ->x, pXYZ->y, pXYZ->z);
    }

Public void showDouble
(
char *pName,
double a
)
    {
    printf ("    double %s = %22.14le\n",
            pName, a);
    }

void showInt
(
char *pName,
int a
)
    {
    printf ("    int %s = %d\n",
            pName, a);
    }

void showDoubleDelta
(
char *pName,
double a,
double aRef
)
    {
    printf ("    double %s = %22.14lf, %10.14lf\n",
            pName, a, a-aRef);
    }

/*---------------------------------------------------------------------------------**//**
* @bsimethod                                    Earlin.Lutz                     12/2005
+---------------+---------------+---------------+---------------+---------------+------*/
static void    roundTripDPoint3dToFloat
(
DPoint3d *pOut,
DPoint3d const *pIn,
void *pContext
)
    {
    float a;
    a = (float)pIn->x;
    pOut->x = (double)a;
    a = (float)pIn->y;
    pOut->y = (double)a;
    a = (float)pIn->z;
    pOut->z = (double)a;
    }

/*---------------------------------------------------------------------------------**//**
* @bsimethod                                    Earlin.Lutz                     12/2005
+---------------+---------------+---------------+---------------+---------------+------*/
static void    truncateDPoint3d001
(
DPoint3d *pOut,
DPoint3d const *pIn,
void *pContext
)
    {
    double factor = 1000;
    int a;

    a = (int) (pIn->x * factor);
    pOut->x = ((double)a)/factor;
    a = (int) (pIn->y * factor);
    pOut->y = ((double)a)/factor;
    a = (int) (pIn->z * factor);
    pOut->z = ((double)a)/factor;
    }


void testCircle3pts (char *pName, double ax, double bx, double cx, double ay, double by, double cy,
double dx,
double dy
)
    {
    DPoint3d pointA, pointB, pointC, pointB1, pointD;
    DPoint3d trueCenter, bestCenter;
    bsiDPoint3d_setXYZ (&pointA, ax, ay, 0.0);
    bsiDPoint3d_setXYZ (&pointB, bx, by, 0.0);
    bsiDPoint3d_setXYZ (&pointC, cx, cy, 0.0);
    bsiDPoint3d_setXYZ (&pointD, dx, dy, 0.0);
    DEllipse3d ellipseABC;

    bsiDEllipse3d_initFrom3DPoint3dOnArc (&ellipseABC, &pointA, &pointB, &pointC);
    printf ("// %sABC\n", pName);
    showDPoint3d ("pointA", &pointA);
    showDPoint3d ("pointB", &pointB);
    showDPoint3d ("pointC", &pointC);

    DPoint3d pointATRUNC, pointBTRUNC, pointCTRUNC;
    for (int numTest = 100; numTest < 12000; numTest *= 10)
        {
        printf ("\n FLOAT ");
        double centerTolFloat = 1.0e-6 * bsiDPoint3d_magnitude (&ellipseABC.vector0);
        double centerTol001 = 0.001;
        double tolMultiplier = 10.0;
        searchTruncatedCoordinateMidpointsToMatchCenter
                 (
                 &pointATRUNC, &pointBTRUNC, &pointCTRUNC,
                 &trueCenter, &bestCenter,
                 &pointA, &pointB, &pointC,
                 roundTripDPoint3dToFloat, NULL,
                 0.49, 0.51, numTest + 1, centerTolFloat * tolMultiplier
                 );
        printf ("\n 001 ");
        searchTruncatedCoordinateMidpointsToMatchCenter
                 (
                 &pointATRUNC, &pointBTRUNC, &pointCTRUNC,
                 &trueCenter, &bestCenter,
                 &pointA, &pointB, &pointC,
                 truncateDPoint3d001, NULL,
                 0.49, 0.51, numTest + 1, centerTol001 * tolMultiplier
                 );
        }




    showDPoint3d ("centerABC", &ellipseABC.center);

    bsiDEllipse3d_fractionParameterToDPoint3d (&ellipseABC, &pointB1, 0.5);
    DEllipse3d ellipse1, ellipseADC;

    bsiDEllipse3d_initFrom3DPoint3dOnArc (&ellipse1, &pointA, &pointB1, &pointC);
    printf ("// %sAB1C\n", pName);
    showDPoint3d ("pointB1", &pointB1);
    showDPoint3d ("centerAB1C", &ellipse1.center);

    bsiDEllipse3d_initFrom3DPoint3dOnArc (&ellipseADC, &pointA, &pointD, &pointC);
    printf ("// %sADC\n", pName);
    showDPoint3d ("pointD", &pointD);
    showDPoint3d ("centerADC", &ellipseADC.center);

    double rABC = bsiDPoint3d_magnitude (&ellipseABC.vector0);
    showDouble      ("Radius_ABC          ", rABC);
    showDoubleDelta ("distance_ABCCenter_B", bsiDPoint3d_distance (&ellipseABC.center, &pointD), rABC);
    showDoubleDelta ("Radius_AB1C         ", bsiDPoint3d_magnitude (&ellipse1.vector0), rABC);
    showDoubleDelta ("distance_ABCCenter_D", bsiDPoint3d_distance (&ellipseABC.center, &pointD), rABC);
    showDoubleDelta ("distance_ADCCenter_B", bsiDPoint3d_distance (&ellipseADC.center, &pointB), rABC);
    showDoubleDelta ("Radius_ADC          ", bsiDPoint3d_magnitude (&ellipseADC.vector0), rABC);

    showDouble (" Sweep_ABC ", ellipseABC.sweep);
    double t = tan (0.25 * ellipseABC.sweep);
    showDouble (" dRadius_dMidCoord", 1.0 / (t*t));
    }

int main(int argc, char * argv[])
    {
    initErrorTracking (NULL, argc, argv);

    //printStandardViews ();
    testCircle3pts ("MS1", 155768.887, 156068.113, 155967.511, 387582.038, 387713.594, 387889.003,
                    156038.966, 387625.025);

    testCircle3pts ("MS2", 155562.525, 155964.932, 156284.794, 387685.22, 387961.23, 388167.592,
                    155920.254, 387931.504);

    testCircle3pts ("MS3", 156186.771, 156393.134, 156692.36, 387899.321, 388064.411, 388131.479,
                    156412.738, 388073.823);

    testCircle3pts ("MS2", 155562.52527255388, 155964.9325285443,
                    156284.794458254346, 387685.22032018349,
                    387961.237298129088750, 388167.5928323426192777,
                    155920.254, 387931.504);


    return getExitStatus();
    }
