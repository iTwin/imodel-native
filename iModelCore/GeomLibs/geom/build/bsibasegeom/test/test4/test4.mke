#--------------------------------------------------------------------------------------
#
#  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
#
#--------------------------------------------------------------------------------------
TESTAPP_NAME=test4

dependencies = $(_MakeFilePath)ellipsefixup.fdf $(_MakeFilePath)ellipsefixup.cpp

%include    $(_MakeFilePath)/../testapp.mki
