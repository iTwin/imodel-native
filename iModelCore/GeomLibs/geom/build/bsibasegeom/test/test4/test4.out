 geomlibs test environment -- command line arguments:
 Tail of argv[0] = bsigeomtest4.exe
// MS1ABC
    DPoint3d pointA = {155768.88699999999, 387582.03800000000,      0.00000000000};
    DPoint3d pointB = {156068.11300000001, 387713.59399999998,      0.00000000000};
    DPoint3d pointC = {155967.51100000000, 387889.00300000003,      0.00000000000};

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  2.91038304567337e-011
    double bestDist =  2.91038304567337e-011
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.83852563232356e-003
    int maxTest = 101
    int numTest = 0

 001     double centerShift =  7.99023905063601e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  3.03431721438851e-004
    double bestDist =  3.03431721438851e-004
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 101
    int numTest = 0

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  2.91038304567337e-011
    double bestDist =  2.91038304567337e-011
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.83852563232356e-003
    int maxTest = 1001
    int numTest = 0

 001     double centerShift =  7.99023905063601e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  3.03431721438851e-004
    double bestDist =  3.03431721438851e-004
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 1001
    int numTest = 0

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  2.91038304567337e-011
    double bestDist =  2.91038304567337e-011
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.83852563232356e-003
    int maxTest = 10001
    int numTest = 0

 001     double centerShift =  7.99023905063601e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  3.03431721438851e-004
    double bestDist =  3.03431721438851e-004
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 10001
    int numTest = 0
    DPoint3d centerABC = {155884.60855549388, 387724.90257433942,      0.00000000000};
// MS1AB1C
    DPoint3d pointB1 = {156038.96577533722, 387625.02458032005,      0.00000000000};
    DPoint3d centerAB1C = {155884.60855549385, 387724.90257433942,      0.00000000000};
// MS1ADC
    DPoint3d pointD = {156038.96599999999, 387625.02500000002,      0.00000000000};
    DPoint3d centerADC = {155884.60852561588, 387724.90259367216,      0.00000000000};
    double Radius_ABC           =  1.83852563232356e+002
    double distance_ABCCenter_B =     183.85252386186477, -0.00003937049124
    double Radius_AB1C          =     183.85256323233767, -0.00000000001833
    double distance_ABCCenter_D =     183.85252386186477, -0.00003937049124
    double distance_ADCCenter_B =     183.85259424288122, 0.00003101052522
    double Radius_ADC           =     183.85255944909088, -0.00000378326513
    double  Sweep_ABC  =  3.35461311088656e+000
    double  dRadius_dMidCoord =  8.07813253351373e-001
// MS2ABC
    DPoint3d pointA = {155562.52499999999, 387685.21999999997,      0.00000000000};
    DPoint3d pointB = {155964.93200000000, 387961.22999999998,      0.00000000000};
    DPoint3d pointC = {156284.79399999999, 388167.59200000000,      0.00000000000};

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  9.65082535221805e-008
    double bestDist =  9.65082535221805e-008
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.53833091203042e-001
    int maxTest = 101
    int numTest = 0

 001     double centerShift =  6.97504793316081e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  1.30672206370071e-001
    double bestDist =  5.73114149653935e-003
    double ratio =  4.38589173301966e-002
    double bestFraction =  5.02000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 101
    int numTest = 61

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  9.65082535221805e-008
    double bestDist =  9.65082535221805e-008
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.53833091203042e-001
    int maxTest = 1001
    int numTest = 0

 001     double centerShift =  6.97504793316081e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  1.30672206370071e-001
    double bestDist =  3.59116718901680e-003
    double ratio =  2.74822572356849e-002
    double bestFraction =  4.90100000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 1001
    int numTest = 6

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  9.65082535221805e-008
    double bestDist =  9.65082535221805e-008
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.53833091203042e-001
    int maxTest = 10001
    int numTest = 0

 001     double centerShift =  6.97504793316081e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  1.30672206370071e-001
    double bestDist =  3.59116718901680e-003
    double ratio =  2.74822572356849e-002
    double bestFraction =  4.90100000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 10001
    int numTest = 51
    DPoint3d centerABC = {164463.91333512671, 375138.84661172138,      0.00000000000};
// MS2AB1C
    DPoint3d pointB1 = {155920.25450210951, 387931.50439796125,      0.00000000000};
    DPoint3d centerAB1C = {164463.91333507313, 375138.84661180165,      0.00000000000};
// MS2ADC
    DPoint3d pointD = {155920.25399999999, 387931.50400000002,      0.00000000000};
    DPoint3d centerADC = {164463.98590850749, 375138.73794558167,      0.00000000000};
    double Radius_ABC           =  1.53833091203042e+004
    double distance_ABCCenter_B =   15383.30906822644100, -0.00005207778122
    double Radius_AB1C          =   15383.30912020775400, -0.00000009646828
    double distance_ABCCenter_D =   15383.30906822644100, -0.00005207778122
    double distance_ADCCenter_B =   15383.43979171554200, 0.13067141131978
    double Radius_ADC           =   15383.43974043281200, 0.13062012858973
    double  Sweep_ABC  =  5.64671531613337e-002
    double  dRadius_dMidCoord =  5.01730491991129e+003
// MS3ABC
    DPoint3d pointA = {156186.77100000001, 387899.32100000000,      0.00000000000};
    DPoint3d pointB = {156393.13399999999, 388064.41100000002,      0.00000000000};
    DPoint3d pointC = {156692.35999999999, 388131.47899999999,      0.00000000000};

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  6.50781433068853e-011
    double bestDist =  6.50781433068853e-011
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  6.33959158852640e-003
    int maxTest = 101
    int numTest = 0

 001     double centerShift =  1.28959102375890e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  2.39734600825543e-003
    double bestDist =  2.39734600825543e-003
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 101
    int numTest = 0

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  6.50781433068853e-011
    double bestDist =  6.50781433068853e-011
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  6.33959158852640e-003
    int maxTest = 1001
    int numTest = 0

 001     double centerShift =  1.28959102375890e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  2.39734600825543e-003
    double bestDist =  2.39734600825543e-003
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 1001
    int numTest = 0

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  6.50781433068853e-011
    double bestDist =  6.50781433068853e-011
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  6.33959158852640e-003
    int maxTest = 10001
    int numTest = 0

 001     double centerShift =  1.28959102375890e-004
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  2.39734600825543e-003
    double bestDist =  2.39734600825543e-003
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 10001
    int numTest = 0
    DPoint3d centerABC = {156677.28508170543, 387497.69909977313,      0.00000000000};
// MS3AB1C
    DPoint3d pointB1 = {156412.73842463488, 388073.82346249872,      0.00000000000};
    DPoint3d centerAB1C = {156677.28508170540, 387497.69909977319,      0.00000000000};
// MS3ADC
    DPoint3d pointD = {156412.73800000001, 388073.82299999997,      0.00000000000};
    DPoint3d centerADC = {156677.28608210091, 387497.69692113210,      0.00000000000};
    double Radius_ABC           =  6.33959158852640e+002
    double distance_ABCCenter_B =     633.95891574410336, -0.00024310853701
    double Radius_AB1C          =     633.95915885258091, -0.00000000005946
    double distance_ABCCenter_D =     633.95891574410336, -0.00024310853701
    double distance_ADCCenter_B =     633.96155478814876, 0.00239593550839
    double Radius_ADC           =     633.96131309011423, 0.00215423747386
    double  Sweep_ABC  =  9.08491339528323e-001
    double  dRadius_dMidCoord =  1.87223628345641e+001
// MS2ABC
    DPoint3d pointA = {155562.52527255387, 387685.22032018349,      0.00000000000};
    DPoint3d pointB = {155964.93252854430, 387961.23729812907,      0.00000000000};
    DPoint3d pointC = {156284.79445825436, 388167.59283234261,      0.00000000000};

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  8.23501252846323e-008
    double bestDist =  8.23501252846323e-008
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.53693472621955e-001
    int maxTest = 101
    int numTest = 0

 001     double centerShift =  6.03468448272777e-004
    double startShift =  4.20479612211658e-004
    double endShift =  9.50153285533091e-004
    double baseDist =  5.01646343387973e-001
    double bestDist =  7.49428580579487e-003
    double ratio =  1.49393809096278e-002
    double bestFraction =  5.07800000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 101
    int numTest = 90

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  8.23501252846323e-008
    double bestDist =  8.23501252846323e-008
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.53693472621955e-001
    int maxTest = 1001
    int numTest = 0

 001     double centerShift =  6.03468448272777e-004
    double startShift =  4.20479612211658e-004
    double endShift =  9.50153285533091e-004
    double baseDist =  5.01646343387973e-001
    double bestDist =  8.20591823569351e-003
    double ratio =  1.63579747841340e-002
    double bestFraction =  4.94380000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 1001
    int numTest = 220

 FLOAT     double centerShift =  0.00000000000000e+000
    double startShift =  0.00000000000000e+000
    double endShift =  0.00000000000000e+000
    double baseDist =  8.23501252846323e-008
    double bestDist =  8.23501252846323e-008
    double ratio =  1.00000000000000e+000
    double bestFraction =  5.00000000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.53693472621955e-001
    int maxTest = 10001
    int numTest = 0

 001     double centerShift =  6.03468448272777e-004
    double startShift =  4.20479612211658e-004
    double endShift =  9.50153285533091e-004
    double baseDist =  5.01646343387973e-001
    double bestDist =  6.38827298263923e-003
    double ratio =  1.27346148673081e-002
    double bestFraction =  4.90082000000000e-001
    double f0 =  5.10000000000000e-001
    double f1 =  4.90000000000000e-001
    double centerTol =  1.00000000000000e-002
    int maxTest = 10001
    int numTest = 42
    DPoint3d centerABC = {164456.16114428596, 375150.46558596566,      0.00000000000};
// MS2AB1C
    DPoint3d pointB1 = {155920.25176776128, 387931.50961146347,      0.00000000000};
    DPoint3d centerAB1C = {164456.16114433174, 375150.46558589721,      0.00000000000};
// MS2ADC
    DPoint3d pointD = {155920.25399999999, 387931.50400000002,      0.00000000000};
    DPoint3d centerADC = {164464.38468656514, 375138.15225749876,      0.00000000000};
    double Radius_ABC           =  1.53693472621955e+004
    double distance_ABCCenter_B =   15369.34135598835700, -0.00590620713228
    double Radius_AB1C          =   15369.34726227780800, 0.00000008231837
    double distance_ABCCenter_D =   15369.34135598835700, -0.00590620713228
    double distance_ADCCenter_B =   15384.15408612777400, 14.80682393228381
    double Radius_ADC           =   15384.14827010141200, 14.80100790592223
    double  Sweep_ABC  =  5.65184914030227e-002
    double  dRadius_dMidCoord =  5.00819296984574e+003
Exit with 0 errors, 0 OK
